### PR TITLE
Add MS COCO dataset support (closes #137)

### LINKED ISSUE #137: [Dataset] Add MS Coco Datasets for image segmentation
Please add the following [MS Coco Detection](https://cocodataset.org/#detection-2016) Dataset as new _Image detection or segmentation_ datasets.

### PR DESCRIPTION
**Pull request by koshtiakanksha**
This PR adds the MS COCO dataset as requested in Issue #137.
 
It includes:
- `coco_dataset()` implementation for detection/segmentation.
- Dataset initialization, structure, and basic tests.

### CODE REVIEW

File: R/coco-dataset.R [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#' MS COCO Dataset
+#'
+#' Loads MS COCO Detection Dataset for object detection and segmentation.
+#'
+#' @param root Root directory where data is downloaded or stored.
+#' @param split One of "train", "val".
```
[the remark this answers is not in the document]
Comment (koshtiakanksha) [reply R/coco-dataset.R]: Thanks for the suggestion. I’ve updated the interface to use train = TRUE following the pattern used in CIFAR and MNIST.

File: R/coco-dataset.R [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#' @param year Dataset year, e.g., "2017".
+#' @param download Whether to download the dataset.
+#' @param ... Additional arguments passed to the dataset.
+#'
+#' @export
+coco_dataset <- dataset(
+  "coco",
+  initialize = function(root, split = "train", year = "2017", download = FALSE, ...) {
```
[the remark this answers is not in the document]
Comment (koshtiakanksha) [reply R/coco-dataset.R]: I’ve made the valid values for the year parameter explicit by adding year <- match.arg(year, c("2014", "2016", "2017")). This now clearly communicates to users what years are supported.

**Comment by koshtiakanksha** [comment]
Thanks for the review! I’ve addressed all points:

- Replaced split with train = TRUE, matching other datasets like mnist and cifar10.
- Used match.arg(year, c("2014", "2016", "2017")) for clarity on valid year values.
- Removed all TODOs — the dataset is complete with full download, load, and parse logic.
- Added unit tests covering download, loading, field checks, and edge cases.

Let me know if the tests look good or if any changes are needed.

**Comment by koshtiakanksha** [comment]
Hi @cregouby,
I’ve made the requested changes:

-Added structured get_resource_info() and MD5 checking via download_and_cache().
-Updated file name and export (dataset_coco_detection), with roxygen docs.
-Followed the CIFAR-style layout, though not inherited directly, since COCO’s JSON structure is quite different.

However, I’m running into an error when calling dataset_coco_detection$new(...):
object of type 'closure' is not subsettable.
Even after restarting R and running devtools::load_all(), typeof(dataset_coco_detection) still returns "closure" instead of a dataset generator. I’ve double-checked naming, export, and load order, but the issue persists. Any advice would be really helpful.

I know the initial code quality was lacking, and I’m doing my best to improve. I’d sincerely appreciate any feedback to help me grow and make better contributions going forward.

Thanks so much for your time and support.

**Comment by cregouby** [comment]
> Hi @cregouby, I’ve made the requested changes:

> However, I’m running into an error when calling dataset_coco_detection$new(...): object of type 'closure' is not subsettable. Even after restarting R and running devtools::load_all(), typeof(dataset_coco_detection) still returns "closure" instead of a dataset generator. I’ve double-checked naming, export, and load order, but the issue persists. Any advice would be really helpful.

**todo** You are callling `self$check_files()` in line 36, this is not defined. You should add a definition for the function.
**improvement** file is now called `dataset_coco.R` which is what I was asking for, but please keep the function name as it used to be `coco_detection_dataset()`, so that it is consistant with dataset functions. (Sorry I was maybe not clear enough)

File: R/dataset-coco.R [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    root <- fs::path_expand(root)
+    self$root <- root
+    self$year <- year
+    self$split <- split
+    self$transforms <- transforms
+    self$target_transform <- target_transform
+
+    self$data_dir <- fs::path(root, glue::glue("coco{year}"))
+    self$image_dir <- fs::path(self$data_dir, glue::glue("{split}{ifelse(year == '2016', '2014', year)}"))
```
Comment (cregouby) [inline R/dataset-coco.R:29]: **todo** glue cannot handle function execution. Please move the condition handling out of it.

**Review by cregouby — CHANGES_REQUESTED** [review]

**Comment by koshtiakanksha** [comment]
Hi @cregouby,
All comments have been addressed, including the switch to jpeg::readJPEG, check_files() addition, and fixing the glue condition. Image loading now works correctly and tests pass.
Please let me know if anything else needs adjustment. Thank you!

File: R/dataset-coco.R [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    if (download)
+      self$download()
+
+    if (!self$check_files())
+      rlang::abort("Dataset files not found. Use download = TRUE to fetch them.")
+
+    self$load_annotations()
+  },
+
+  check_files = function() {
```
Comment (cregouby) [inline R/dataset-coco.R:50]:
```suggestion
    if (!self$check_exists())
      runtime_error("Dataset not found. You can use `download = TRUE` to download it.")

    self$load_annotations()
  },

  check_exists = function() {
```

File: R/dataset-coco.R [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    target <- list(
+      image_id = image_id,
+      boxes = boxes,
+      labels = labels,
+      area = area,
+      iscrowd = iscrowd,
+      segmentation = segmentation,
+      height = info$height,
+      width = info$width
```
Comment (cregouby) [inline R/dataset-coco.R:85]: **question** what is it for  ? do we realy want a model to infer on image shape ?

File: R/dataset-coco.R [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+#' Loads MS COCO Dataset for object detection and segmentation.
+#'
+#' @param root Root directory for data.
+#' @param train Whether to load training split (TRUE) or validation (FALSE).
+#' @param year Dataset year: "2014", "2016", or "2017".
+#' @param download If TRUE, downloads the dataset if needed.
+#' @param transforms Transform applied to image.
+#' @param target_transform Transform applied to target.
+#'
+#' The returned image is in CHW format (channels, height, width), matching torch convention.
```
Comment (cregouby) [inline R/dataset-coco.R:14]: **missing** the complex and unique output names and data-model requires a proper @return section refering to a (required)  details section.
**missing** an example highlighting the outputs, with maybe `draw_segmentation_mask` or `draw_bounding_box`

**Comment by koshtiakanksha** [comment]
Hi @cregouby,
I’ve made the requested changes.

I also want to sincerely apologize that this PR is taking more time and review than expected. I really appreciate your patience and guidance. I assure you that once I get this one right, I’ll have a much better understanding and will be able to contribute more efficiently going forward.

To confirm, I tested the dataset using the following code:

ds <- coco_detection_dataset(root = "~/data", train = FALSE, year = "2017", download = TRUE)
sample <- ds[1]
image <- sample$image

image_uint8 <- image$mul(255)$clamp(0, 255)$to(dtype = torch::torch_uint8())
img_array <- as.array(image_uint8$permute(c(2, 3, 1)))  # CHW -> HWC
img_numeric <- as.numeric(img_array) / 255
dim(img_numeric) <- dim(img_array)

plot(as.raster(img_numeric)) # Plot

image

File: R/dataset-coco.R [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+#' @param target_transform Optional transform function applied to the target (labels, boxes, etc.).
+#'
+#' @return
+#' An R6 dataset object. Each item is a list with two elements:
+#' \describe{
+#'   \item{image}{A 3D \code{torch_tensor} of shape \code{[C, H, W]} (channel-first).}
+#'   \item{target}{A list with:
+#'     \describe{
+#'       \item{boxes}{A matrix of bounding boxes in \code{[x1, y1, x2, y2]} format.}
+#'       \item{labels}{A numeric vector of integer class labels.}
```
Comment (cregouby) [inline R/dataset-coco.R:19]: **todo** "a numeric vector" or "a integer vector", but not both...

File: R/dataset-coco.R [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+  },
+
+  download = function() {
+    info <- self$get_resource_info()
+
+    ann_zip <- download_and_cache(info$ann_url)
+
+    img_zip <- download_and_cache(info$img_url)
+    if (tools::md5sum(img_zip) != info$img_md5) {
+      rlang::abort(glue::glue("Corrupt file! Delete the file in {img_zip} and try again."))
```
Comment (cregouby) [inline R/dataset-coco.R:137]: **todo** the line must match **exactly** the one in other dataset MD5 failure message in order  to be translated See https://github.com/mlverse/torchvision/blob/5b46ee49f7f2c846add8fde8cd51c120d1be406e/R/dataset-fgvc.R#L168. Thus please rename the variable accordingly.

**Review by cregouby — CHANGES_REQUESTED** [review]

File: R/dataset-coco.R [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+#'
+#' # Ensure bounding boxes are torch tensors
+#' boxes <- if (!inherits(target$boxes, "torch_tensor")) {
+#'   torch::torch_tensor(target$boxes, dtype = torch::torch_float())
+#' } else {
+#'   target$boxes
+#' }
+#'
+#' # Convert labels to character
+#' labels <- as.character(torch::as_array(target$labels))
```
Comment (cregouby) [inline R/dataset-coco.R:56]: **improvement** `target$labels` here are factor levels turned into character. We have to find something usable out of `ds$categories` that could be much more expressive for the end-user (plus would allow a visual validation of the correctness of the mapping). 

(My own try for example is wrong : )
image

File: R/dataset-coco.R [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+#'
+#' # Draw bounding boxes
+#' output <- torchvision::draw_bounding_boxes(
```
Comment (cregouby) [inline R/dataset-coco.R:59]: **improvement** we are within torchvision here, so you don't have to mention `torchvision::`

File: R/dataset-coco.R [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+#'   image = image_uint8,
+#'   boxes = boxes,
+#'   labels = labels
+#' )
+#'
+#' # Convert to array and plot
+#' output_array <- as.array(output$permute(c(2, 3, 1)))  # CHW to HWC
+#' output_array <- as.numeric(output_array) / 255
+#' dim(output_array) <- dim(as.array(output$permute(c(2, 3, 1))))
+#' plot(as.raster(output_array))
```
Comment (cregouby) [inline R/dataset-coco.R:69]: **improvement** Won't you prefer the straingtforward one-line code `tensor_image_browse(output)` ?

**Review by cregouby — CHANGES_REQUESTED** [review]
There is still room for example and documentation improvement. See the inline comments.

**Comment by cregouby** [comment]
**suggestion** in order to make the COCO dataset more accessible, we could also manage the coco dataset item as a S3 class that would be directly managed by `draw_bounding_boxes(()` and `draw_segmentation_mask()` . This would turn the example into the very straigthforward

```
coco <- coco_detection_dataset( root = "~/data", train = FALSE, year = "2017", download = TRUE)
item <- coco[1]
segmented_item <- draw_segmentation_mask(item)
tensor_image_browse(segmented_item)
```
What is your view on this ?
